---
title: "sctherapy_tutorial"
output: html_document
date: "2025-02-20"
creater: Kristen Michelle Nader
email: kristen.nader@helsinki.fi
---

This is a tutorial on how to use sctherapy from within the ican platform. The github has already been cloned into this docker (which included R source files and sample data).

This markdown will be a remake of the github tutorial using the docker. 


scTherapy is a computational framework to predict personalized monotherapies and multi-targeting drug combinations for cancer patients using only single-cell transcriptomic data.

The tool consists of two main steps:

A semi-automated approach to categorize cells into healthy(normal) and malignant, followed by the separation of malignant cells into subpopulations/clones (Fig 1, panel a).
Next, it identifies differentially expressed genes between healthy and malignant cells, which are subsequently used as input to our pre-trained machine learning model. The model predicts potential treatment options that effectively target malignant cells while minimizing toxicity to healthy cells (Fig 1, panel b).
For more information, please refer to original nature communication publication.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Step 1.0: Load libraries and process the data

First, let's load all the necessary libraries and source functions required for the analysis.


```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
invisible(lapply(c("dplyr","Seurat","HGNChelper","openxlsx","copykat","copykatRcpp","ggplot2", "SCEVAN","yaGST","cowplot","Rcpp","Rclusterpp",
          "parallel","biomaRt","logger","httr", "jsonlite", "readr","future"), library, character.only = !0))


r_path=  "/home/rstudio/scTherapy/R/"
r_files = list.files(path = r_path, pattern = "\\_v5.R$", full.names = TRUE)
pred_files = list.files(path = r_path, pattern = "\\compounds.R$", full.names = TRUE)
invisible(lapply(r_files, source))
invisible(lapply(pred_files, source))

rm(r_path, r_files, pred_files,read_zip)
gc()


```

You should do a quick sanity check that there are functions in the 'Enviroment' pane.
Next, let's load an example PBMC scRNA-seq dataset, consisting of ~3000 cells obtained from a human AML patient. The count matrix is normalized and clustered using Seurat (see Seurat tutorial for more details). An example raw count matrix data can be found here.

```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
temp_zip="/home/rstudio/scTherapy/sample_x_exp.rawdata.txt.zip"
temp_dir=tempdir()
file_inside_zip <- unzip(temp_zip, list = TRUE)$Name[1]
unzip(temp_zip, files = file_inside_zip, exdir = temp_dir)
data <- suppressWarnings(data.table::fread(file.path(temp_dir, file_inside_zip)) %>% as.data.frame())
rownames(data) <- data[[1]]; data <- data[,-1]

dim(data)

rm(temp_zip,file_inside_zip,temp_dir)
gc()
```

Sanity check: the dimensions of the sample data should be 15219 2162. 

```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
patient_sample <- CreateSeuratObject(counts = data)	
   
# simple filtering
patient_sample[["percent.mt"]] <- PercentageFeatureSet(patient_sample, pattern = "^MT-")
patient_sample <- subset(patient_sample, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

# normalize data
patient_sample <- NormalizeData(patient_sample, normalization.method = "LogNormalize", scale.factor = 10000)
patient_sample <- FindVariableFeatures(patient_sample, selection.method = "vst", nfeatures = 2000)

# scale and PCA
patient_sample <- ScaleData(patient_sample, features = rownames(patient_sample))
patient_sample <- RunPCA(patient_sample, features = VariableFeatures(object = patient_sample))

# check number of PC (optional)
ElbowPlot(patient_sample)

# Cluster and visualize
patient_sample <- FindNeighbors(patient_sample, dims = 1:10)
patient_sample <- FindClusters(patient_sample, resolution = 0.8)
patient_sample <- RunUMAP(patient_sample, dims = 1:10)
DimPlot(patient_sample, reduction = "umap")


rm(data)
gc()
```
Sanity check: the Dimplot should plot 

## Step 1.1: Automated Cell type annotation with ScType
In this step, we use our method for fully-automated cell type annotation called ScType. It only requires an scRNA-seq object and the name of the tissue type as input, please see ScType GitHub for more details: ScType.

For our AML patient sample, we specify known_tissue_type as Immune system, but other possible tissue types include: Immune system, Pancreas, Liver, Eye, Kidney, Brain, Lung, Adrenal, Heart, Intestine, Muscle, Placenta, Spleen, Stomach, Thymus.


```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
patient_sample <- run_sctype(patient_sample,known_tissue_type="Immune system",plot=TRUE)
```

##Step 1.2: Identification of malignant/healthy clusters

In this step, we use an ensemble of three tools (CopyKat, scType+new markers, and SCEVAN) to confidently classify cells into malignant and healthy groups. To enhance accuracy, we recommend providing prior knowledge of a confident healthy cell cluster as input to run_ensemble function.

Here, we provide T cells as confident healthy cell cluster, given that T cells are considered as normal cells in AML - PMID:30827681

```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
norm_cells <- get_normal_cells(patient_sample,c("Memory CD4+ T cells"))
patient_sample <- run_ensemble(patient_sample,disease="AML",known_normal_cells=norm_cells,plot=FALSE)
visualize_ensemble_step(patient_sample)
```
Sanity check: ensemble plot outputs to the screen ( 5 plots A-B-C)

Users have the option to choose between two options: making predictions to target the entire malignant tumor mass or exploring clonal lineages to predict combination therapies aimed at targeting specific subclones.

-A tutorial on prediciting single agent therapies: Predict monotherapies using malignant specific DEG
-A tutorial on prediciting combination therapies: Predict combination therapies using subclone specific DEG. This includes infering clonal architecture within the malignant cluster identified in step 1.2.

#Predict monotherapies using malignant specific DEG

At this point, users can use scTherapy to predict monotherapies to target the malignant cluster identified in step 1.2

##Step 2.1: Extract malignant cluster specific DEG
```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
plan("multisession", workers = 4)
malignant_cells_DEG <- clone_DEG(patient_sample,malignant_identifier="malignant",known_normal_cells="healthy",save=FALSE)
```
First, we begin by filtering the differentially expressed genes based on 2 critera:

## Step 2.2: Use malignant specific DEG as input to the pre-trained LightGBM model.
First, we begin by filtering the differentially expressed genes based on 2 critera:
  1. (p_val_adj <= 0.05 & (avg_log2FC > 1 | avg_log2FC < -1))
  2. (avg_log2FC > -0.1 & avg_log2FC < 0.1)
Then, we can run predict_drugs using the malignant DEGs as input to predict monotherpies.
```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
# filter DEGS
gene_list <- "/home/rstudio/scTherapy/geneinfo_beta_input.txt"
gene_info <- data.table::fread(gene_list) %>% as.data.frame()

DEG_malignant <- malignant_cells_DEG %>%
    mutate(gene_symbol = rownames(.)) %>% inner_join(gene_info, by = "gene_symbol") %>%
    filter((p_val_adj <= 0.05 & (avg_log2FC > 1 | avg_log2FC < -1)) | (avg_log2FC > -0.1 & avg_log2FC < 0.1))
DEG_malignant_list <- setNames(as.list(DEG_malignant$avg_log2FC), DEG_malignant$gene_symbol)
	
#predict monotherapies for malignant cluster
monotherapy_drugs <- predict_drugs(DEG_malignant_list)
```

#Predict combination therapies using subclone specific DEG

If users are interested in exploring tumor subclones and targeting them with specific compounds, they can follow the subsequent steps outlined below.

Note: For samples with a limited number of cancer cells detected in scRNA-seq, we advise to focus solely on predicting responses to individual drugs. Some samples may contain insufficient number of cancer cells to reliably identify clonal subpopulations


##Step 3.1: Identification of genetically distinct subclones
This step uses healthy/reference cells identified by step 1.2(ensemble model) to identify genetically distinct sublcones. Note that this step may be computationally intensive.

```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
# Load extra required libraries 
lapply(c("rjags","biomaRt","infercnv"), library, character.only = !0)
# runs infercnv and all helper functions for the analysis 
patient_sample <- run_infercnv(patient_sample)
```


##Step 3.2: Extract subclone specific DEG
We will focus on broad levels subclones in this tutorial, but more specific subclones can be used in this step for more specific analysis. For subclones A and B:
```{r load, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
plan("multisession", workers = 4)
subcloneA <- clone_DEG(patient_sample,malignant_identifier="A",known_normal_cells="healthy",save=FALSE)
subcloneB <- clone_DEG(patient_sample,malignant_identifier="B",known_normal_cells="healthy",save=FALSE)

```


#Step 3.3: Use subclone specific DEG as input to the pre-trained LightGBM model.
First, we begin by filtering the differentially expressed genes based on 2 critera with :

1. (p_val_adj <= 0.05 & (avg_log2FC > 1 | avg_log2FC < -1))
2. (avg_log2FC > -0.1 & avg_log2FC < 0.1)

```{r}

gene_list <- "/home/rstudio/scTherapy/geneinfo_beta_input.txt"
gene_info <- data.table::fread(gene_list) %>% as.data.frame()

# filter subclone "A" specific DEG					 
DEG_subclone_A <- subcloneA %>%
    mutate(gene_symbol = rownames(.)) %>% inner_join(gene_info, by = "gene_symbol") %>%
    filter((p_val_adj <= 0.05 & (avg_log2FC > 1 | avg_log2FC < -1)) | (avg_log2FC > -0.1 & avg_log2FC < 0.1))
DEG_subclone_A_list <- setNames(as.list(DEG_subclone_A$avg_log2FC), DEG_subclone_A$gene_symbol)      
	
# filter subclone "B" specific DEG	
DEG_subclone_B <- subcloneB %>%
    mutate(gene_symbol = rownames(.)) %>% inner_join(gene_info, by = "gene_symbol") %>%
    filter((p_val_adj <= 0.05 & (avg_log2FC > 1 | avg_log2FC < -1)) | (avg_log2FC > -0.1 & avg_log2FC < 0.1))
DEG_subclone_B_list <- setNames(as.list(DEG_subclone_B$avg_log2FC), DEG_subclone_B$gene_symbol)   
```

For each run of predict_drugs, the model predicts drug:dose based on a predefined set of drug:dose:response integrated from LINCS L1000 and PharmacoDB.

```{r}
# predict subclone A specific drug:dose
subcloneA_drugs <- predict_drugs(degs_list=DEG_subclone_A_list)

# predict subclone B specific drug:dose
subcloneB_drugs <- predict_drugs(degs_list=DEG_subclone_B_list)
```

```{r}
sessionInfo()
```


